### DISCUSSION TITLE
About firmware format and implementation

### DISCUSSION THREAD

**Discussion by randytry** [fwupd/fwupd#5329] [Q&A] [answered]
Hi, I'm developing a custom plugin, and i found some fwupdtool command about firmware file format:
In my application case:
I just defined a new beta firmware format define by RIFF format 
```
typedef struct ChunkHeader
{
	UINT32 FourCC;
	UINT32 Size; // payload size
}Chunk_Header;

typedef struct FlashSpace
{
	UINT32 Address;
	UINT32 Size;
}Flash_Space;


typedef struct FlashBinaryChunk {
	Chunk_Header Header;
        Flash_Space SpaceToErase;
        Flash_Space Binary;
	BYTE*  BinaryData;  //pay load
}FlashBinary_Chunk;
```
And I found a fwupd already supports many firmware format likes ihex, srec ...
This makes me want to switch to the format already used by fwupd,
my question is if i use format already support, i don't need to create fu-xxx.firmware.c  implement 
fwupdtool firmware-export, firmware-parse,  firmware-extract function?

Is there a common firmware Image format recommended?
i just need to package some start address and raw binary, and add ascii header like srec record file s0

**Comment by hughsie** [comment, marked as answer]
We could certainly add support for RIFF (...even if it looks like an audio format!) but you're correct that using an existing format is probably a better idea, and a lot less code. If you use one of the existing formats it'll be a lot less to test too.

I guess the ones that we "recommend" are DfuSe, DFU, Intel Hex, SREC -- although if you need to encode a start address I think that rules out DFU.  If you've got multiple images, or any additional metadata I guess DfuSe makes sense, see https://github.com/fwupd/fwupd/blob/main/libfwupdplugin/tests/dfuse.builder.xml for an example on how it's structured and built. 

On the other hand, creating a simple firmware format isn't the end of the world; fu-xxx.firmware.c only requires ~4 vfuncs to implement and you can make fwupd match what the compiler is producing. What I would also say is that it's worth thinking about SBoMs -- e.g. by including a SBOM section in the DfuSe/SREC/IHEX image or something. e.g. see https://lists.linuxfoundation.org/pipermail/lvfs-announce/2022-October/000074.html for more details. Hope that helps.

**Reply by chenhn123** [reply]
I think DfuSe is good idea for my project for my project, thanks.

**Reply by randytry** [reply]
thanks, dfu document and tool also help me a lot:
http://rc.fdr.hu/UM0391.pdf
https://www.st.com/en/development-tools/stsw-stm32080.html

**Reply by hughsie** [reply]
Cool, I'm glad to help. Also bear in bind you can use fwupd itself to build the DfuSe image if you want, e.g. calling `fwupdtool firmware-build yourfirmware.builder.xml` will build an image from a raw binary and an address.

**Reply by niej** [reply]
Hi @hughsie,
Could you help to elaborate how to build firmware from raw binary? For example, I see below files and ifd.bin is the output of ifd.builder.xml actually. But I cannot find where is the data source in ifd.builder.xml
./libfwupdplugin/tests/ifd.bin  
./libfwupdplugin/tests/ifd.builder.xml

**Reply by hughsie** [reply]
IFD is a container, so doesn't have a direct payload. If you look at something like ihex.builder.xml you can see the data in the `data` tag as a base64 blob. You can also do `filename` with a pointer to the actual payload.

**Reply by niej** [reply]
Thanks! ihex.builder.xml embedded data in it, which is not convenient for large size firmware. While I cannot find any *builder.xml that contains filename key word. Do we have any example to generate with ```fwupdtool firmware-build yourfirmware.builder.xml```?

I find this script which use gcab to generate firmware blob. Do they share the identical functionality? 
plugins/uefi-capsule/tests/example/build.sh  

I guess I am lack of some basic concept of firmware container or type relating to fwupd. Why firmware container/type is needed in fwupd? In my mind, firmware is just a binary that need to be transfer/written to device. So I had assumed that fwupdtool or gcab just wraps the firmware into a archive file, and fwupd can extract the firmware binary and write it to device. Please help to correct me on this, thanks!

**Reply by hughsie** [reply]
Aha, so maybe i confused you. The "payload" is specified in the cabinet archive, so you might have:

    payload.bin (a DfuSe file)
    firmware.metainfo.xml (the metadata)

and that would be created as `whatever.cab` and uploaded to the LVFS. You only need to use `fwupdtool firmware-build` if you're trying to actually build the `payload.bin` file from something not supported by fwupd -- and it's usually better if your compiler/linker outputs the correct file format anyway.

Can you share what file type you have now?

**Reply by niej** [reply]
I have 2 payloads, an openBMC firmware to write to 1st MTD, and host BIOS to write to 2nd MTD. Seems I do not need to care ```fwupdtool firmware-build```
Both are to be updated in BMC system. So I am not sure whether both are MTD devices, or one MTD device and one UEFI device in fwupd domain. Could you help explain that, or point to any doc? Thanks!

**Comment by randytry** [comment]
I found fuse just used by ST products , I am worried about the license issue.
Will there be any costs if used for commercial purposes?

**Reply by hughsie** [reply]
I'm not aware of any; I implemented the DfuSe write and read support in fwupd without signing an NDA.
